doc/port-fec/fec-auto-determination.md

```diff
@@ -0,0 +1,185 @@
+    - Some platforms have default FEC defined in `port_config.ini`, which however might not be suitable for the specific port/optics on the system.
+
+The feature in this document is to address the issue in both of above scenarios in a common platform-independent way, since the rule to determine FEC for a given port/optics is common for all platforms.
+
+### High-Level Design
+
+- Add `determine-fec` module which can determine FEC mode based on common rule for a given port with a given optics.
+    - This module provides a `determine_fec` API which can be invoked by below entities.
+- Enhance today's DPB CLI to automatically determine and configure FEC in CONFIG_DB for newly created ports, based on determine-fec module.
+- Add a user-triggered CLI `fec-auto-correct` to automatically determine and configure FEC in CONFIG_DB for existing ports, based on determine-fec module.
+
+### API design
+```
+def determine_fec(lane_speed: int, num_lanes: int, optics_type: Optional[str] = None) -> str:
+    """
+    Determines the appropriate Forward Error Correction (FEC) type based on lane speed, number of lanes, and optics type for a specific port.
+    This logic is based on FEC mapping rules common for all platforms.
+
+    Parameters:
+    - lane_speed (int): The speed of each lane in GB.
+    - num_lanes (int): The total number of lanes.
+    - optics_type (Optional[str]): The type of optics in use. Can be None if not applicable.
+
+    Returns:
+    - str: The recommended FEC type based on the common rules. It can be either 'none'/'rs'/'fc'.
+
+    Example:
+    >>> determine_fec(25, 4, "100G-SR4")
+    "rs"
+
+    """
+```
+
+### Common Rule for FEC Determination
+
+#### Table 1: FEC Mapping Based on Optics Type
+| Optics Type | FEC  |
+|-------------|------|
+| 40G         | none |
+| 100G-DR     | none |
+| 100G-FR     | none |
+| 100G-LR     | none |
+| 100G-LR4    | none |
+| 100G-ER4    | none |
+| 100G AOC    | none |
+| 400G        | rs   |
+| ALL_OTHER   | rs   |
+
+
+#### Table 2: FEC Mapping Based on Lane Speed and Number of Lanes
+| Lane Speed | Number of Lanes | FEC  |
+|------------|-----------------|------|
+| 10         | 4               | none |
+| 20         | 2               | none |
+| 25         | 2               | rs   |
+| 25         | 4               | rs   |
+| 25         | 8               | rs   |
+| 50         | 1               | rs   |
+| 50         | 2               | rs   |
+| 50         | 4               | rs   |
+| 50         | 8               | rs   |
+| 50         | 16              | rs   |
+
+> [!NOTE]
+> If a port has matched FEC entry in both above tables, then prefers FEC entry in first table.
+> For example: A port with 100G-DR optics running in non-breakout mode (`optics_type=100G-DR, lane_speed=25, num_lane=4`) will have a match in both rules, the matched FEC entry in table 1 will be preferably choosen, which is ```none``` in this case.
```

Comment:
Could we also have the following corner included in the HLD?
i.e.
1. 100G DAC + 4x25G port breakout:  fec=none
2. 100G DAC + 2x50G port breakout:  fec=rs
3. 100G DAC + 1x100G port breakout:  fec=rs

Reply:
1x100G(optics_type=ALL_OTHER, lane_speed=25, num_lanes=4) and 2x50G(optics_type=ALL_OTHER, lane_speed=50, num_lanes=1) is already covered. Will add for others.